draft-ietf-quic-version-negotiation.md

```diff
@@ -67,105 +67,171 @@ Discussion of this work is encouraged to happen on the QUIC IETF mailing list
-provided in its Initial. Version negotiation takes place after the retry cycle
-is over.
-
-In order for negotiation to complete successfully, the client's Initial packet
-(and initial CRYPTO frames) MUST be interpretable by the server. This implies
-that servers must retain the ability to process the Initial packet from older
-versions as long as they are reasonably popular.  This is not generally an issue
-in practice as long as the overall structure of the protocol remains similar.
-
-# Version Negotiation Transport Parameter
-
-This document registers a new transport parameter, `version_negotiation`. The
-contents of this transport parameter depend on whether the client or server is
-sending it, and are shown below:
+This document specifies two means of performing version negotiation: one
+"incompatible" which requires a round trip and is applicable to all versions,
+and one "compatible" that allows saving the round trip but only applies when
+the versions are compatible.
+
+The client initiates a QUIC connection by sending a first flight of QUIC
+packets with a long header to the server {{INV}}. We'll refer to the version of
+those packets as the "original version". The client's first flight includes
+handshake version information (see {{hs-vers-info}}) which will be used to
+optionally enable compatible version negotation (see {{compat-vn}}), and to
+prevent version downgrade attacks (see {{downgrade}}).
+
+Upon receiving this first flight, the server verifies whether it knows how to
+parse first flights from the original version. If it does not, then it starts
+incompatible version negotiation, see {{incompat-vn}}. If the server can parse
+the first flight, it can either establish the connection using the original
+version, or it MAY attempt compatible version negotiation, see {{compat-vn}}.
+
+Note that it is possible for a server to have the ability to parse the first
+flight of a given version without fully supporting it, in the sense that it
+implements enough of the version's specification to parse first flight packets
+but not enough to fully establish a connection using that version.
+
+
+## Incompatible Version Negotiation {#incompat-vn}
+
+The server starts incompatible version negotiation by sending a VN packet,
+listing all the versions that it does support.
+
+Upon receiving the VN packet, the client will search for a version it supports
+in the list provided by the server. If it doesn't find one, it aborts the
+connection attempt. Otherwise, it selects a mutually supported version and
+sends a new first flight with that version - we refer to this version as the
+"negotiated version".
+
+When the client sends its new first flight, it SHOULD use the same 5-tuple
+(source and destination IP addresses and UDP port numbers) and the same
+Destination Connection ID as its original first flight. This will increase the
+likelihood of both first flights reaching the same server.
```

Comment:
This SHOULD seems like a problem. If this is the same connection, then this should be a MUST. If not, then DCID should change.

Reply:
Do you have thoughts on what you prefer? I dislike MUST because it can cause issues if version A allows CID lengths of [0-8] and version B only allows [9-20]. (I know this probably won't happen, but I don't want to paint us into that corner)